src/sensors/softfusion/drivers/icm42688.h

```diff
@@ -170,23 +211,47 @@ struct ICM42688 {
 				&read_buffer[i + 0x1],
 				sizeof(FifoEntryAligned)
 			);  // skip fifo header
-
+#if debug_icm42688_20bit
+			// 6.1 Packet Structure for high resolution mode
+			// https://invensense.tdk.com/wp-content/uploads/2020/04/ds-000347_icm-42688-p-datasheet.pdf
+			// When 20-bits data format is used, gyroscope data consists of 19-bits of
+			// actual data and the LSB is always set to 0
 			const int32_t gyroData[3]{
-				static_cast<int32_t>(entry.part.gyro[0]) << 4 | (entry.part.xlsb & 0xf),
-				static_cast<int32_t>(entry.part.gyro[1]) << 4 | (entry.part.ylsb & 0xf),
-				static_cast<int32_t>(entry.part.gyro[2]) << 4 | (entry.part.zlsb & 0xf),
+				static_cast<int32_t>(entry.part.gyro[0]) << 3
+					| ((entry.part.xlsb & 0xe) >> 1),
+				static_cast<int32_t>(entry.part.gyro[1]) << 3
+					| ((entry.part.ylsb & 0xe) >> 1),
+				static_cast<int32_t>(entry.part.gyro[2]) << 3
+					| ((entry.part.zlsb & 0xe) >> 1),
 			};
+#else
+			const int32_t gyroData[3]{
+				static_cast<int32_t>(entry.part.gyro[0]),
+				static_cast<int32_t>(entry.part.gyro[1]),
+				static_cast<int32_t>(entry.part.gyro[2]),
+			};
+#endif
 			callbacks.processGyroSample(gyroData, GyrTs);
 
 			if (entry.part.accel[0] != -32768) {
+#if debug_icm42688_20bit
+				// accelerometer data consists of 18-bits of actual data and the two
+				// lowest order bits are always set to 0
 				const int32_t accelData[3]{
-					static_cast<int32_t>(entry.part.accel[0]) << 4
-						| (static_cast<int32_t>(entry.part.xlsb) & 0xf0 >> 4),
-					static_cast<int32_t>(entry.part.accel[1]) << 4
-						| (static_cast<int32_t>(entry.part.ylsb) & 0xf0 >> 4),
-					static_cast<int32_t>(entry.part.accel[2]) << 4
-						| (static_cast<int32_t>(entry.part.zlsb) & 0xf0 >> 4),
+					static_cast<int32_t>(entry.part.accel[0]) << 2
+						| (static_cast<int32_t>(entry.part.xlsb) & 0xf0 >> 6),
+					static_cast<int32_t>(entry.part.accel[1]) << 2
+						| (static_cast<int32_t>(entry.part.ylsb) & 0xf0 >> 6),
+					static_cast<int32_t>(entry.part.accel[2]) << 2
+						| (static_cast<int32_t>(entry.part.zlsb) & 0xf0 >> 6),
```

Comment:
`>>` has higher precedence than `&`, this left right shifts the 0xf0 constant first then applies that as a mask.